cmd/mungedocs/util.go

```diff
@@ -112,72 +168,118 @@ func endMungeTag(desc string) string {
 	return fmt.Sprintf("<!-- END MUNGE: %s -->", desc)
 }
 
-// Calls 'replace' for all sections of the document not in ``` / ``` blocks. So
-// that you don't have false positives inside those blocks.
-func replaceNonPreformatted(input []byte, replace func([]byte) []byte) []byte {
-	f := splitByPreformatted(input)
-	output := []byte(nil)
-	for _, block := range f {
-		if block.preformatted {
-			output = append(output, block.data...)
-		} else {
-			output = append(output, replace(block.data)...)
+type mungeLine struct {
+	data         string
+	preformatted bool
+	header       bool
+	link         bool
+	beginTag     bool
+	endTag       bool
+}
+
+type mungeLines []mungeLine
+
+func (m1 mungeLines) Equal(m2 mungeLines) bool {
+	if len(m1) != len(m2) {
+		return false
+	}
+	for i, _ := range m1 {
+		if m1[i].data != m2[i].data {
+			return false
 		}
 	}
-	return output
+	return true
 }
 
-type fileBlock struct {
-	preformatted bool
-	data         []byte
+func (mlines mungeLines) Bytes() []byte {
+	slice := []string{}
+	for _, mline := range mlines {
+		slice = append(slice, mline.data)
+	}
+	s := strings.Join(slice, "\n")
+	b := []byte(s)
+	// Apparently we need to tack on an extra newline at the end of the file
+	b = append(b, '\n')
+	return b
 }
 
-type fileBlocks []fileBlock
-
 var (
 	// Finds all preformatted block start/stops.
 	preformatRE    = regexp.MustCompile("^\\s*```")
 	notPreformatRE = regexp.MustCompile("^\\s*```.*```")
+	// Is this line a header?
+	mlHeaderRE = regexp.MustCompile(`^#`)
+	// Is there a link on this line?
+	mlLinkRE   = regexp.MustCompile(`\[[^]]*\]\([^)]*\)`)
+	beginTagRE = regexp.MustCompile(`<!-- BEGIN MUNGE:`)
+	endTagRE   = regexp.MustCompile(`<!-- END MUNGE:`)
+
+	blankMungeLine = newMungeLine("")
 )
 
-func splitByPreformatted(input []byte) fileBlocks {
-	f := fileBlocks{}
+// Does not set 'preformatted'
+func newMungeLine(line string) mungeLine {
+	ml := mungeLine{
+		data: line,
+	}
+	match := mlHeaderRE.FindString(line)
+	if match != "" {
+		ml.header = true
+	}
+
+	match = mlLinkRE.FindString(line)
+	if match != "" {
+		ml.link = true
+	}
+
+	match = beginTagRE.FindString(line)
+	if match != "" {
+		ml.beginTag = true
+	}
+
+	match = endTagRE.FindString(line)
+	if match != "" {
+		ml.endTag = true
+	}
+	return ml
+}
```

Comment:
maybe...

```go
func newMungeLine(line []byte) mungeLine {
	return mungeLine{
		data:     line,
		header:   mlHeaderRE.Match(line),
		link:     mlLinkRE.Match(line),
		beginTag: beginTagRE.Match(line),
		endTag:   endTag.Match(line),
	}
}
```

There's a MatchString variant if you want to stick with strings.